service/dap/server.go

```diff
@@ -2861,6 +2904,46 @@ func processExited(state *api.DebuggerState, err error) bool {
 	_, isexited := err.(proc.ErrProcessExited)
 	return isexited || err == nil && state.Exited
 }
+func (s *Server) setRunning(running bool) {
+	s.runningMu.Lock()
+	s.running = running
+	s.runningMu.Unlock()
+}
+
+func (s *Server) isRunning() bool {
+	s.runningMu.Lock()
+	defer s.runningMu.Unlock()
+	return s.running
+}
+
+// resume is a helper function to resume the execution
+// of the target when the program is halted, but no
+// stopped event has been sent.
+func (s *Server) resume() (*api.DebuggerState, error) {
+	// No other goroutines should be able to try to resume
+	// or halt execution while this goroutine is continuing
+	// the program.
+	// Hold onto resumeMu until the program is running.
+	resumeNotify := make(chan struct{}, 1)
+	defer s.asyncCommandDone(resumeNotify)
```

Comment:
why is this defer in two different functions?

Reply:
This is closing two different channels. I think I will rename asyncCommandDone to maybeCloseChan or something, since that is what it is being used for.